nft-client/src/main.rs

```diff
@@ -0,0 +1,635 @@
+//! This is an example of an off-chain application which integrates with a smart
+//! contract on chain. The application is a client for working with a specific
+//! implementation of a Non-fungible token(NFT) smart contract.
+//! The smart contract should follow the CTS1 specification, but some additional
+//! assumptions are made for how the contract serializes the state and the
+//! existence of a "mint" contract function.
+//!
+//! The smart contract is called "CTS1-NFT" and can be found [here](https://github.com/Concordium/concordium-rust-smart-contracts/tree/main/examples).
+//!
+//! Most of the functionality needs access to the GRPC API of a running
+//! concordium-node. Some functionality of the client additionally depends on
+//! the concordium-node to have the transaction logging enabled and access to
+//! the PostgreSQL database with transaction logs.
+
+use anyhow::*;
+use clap::AppSettings;
+use concordium_rust_sdk::{
+    common, constants,
+    endpoints::{Client, Endpoint},
+    id, postgres,
+    postgres::DatabaseSummaryEntry,
+    types,
+};
+use futures::{StreamExt, TryStreamExt};
+use serde::*;
+use smart_contracts::concordium_contracts_common;
+use std::{
+    collections::{BTreeMap as Map, BTreeSet as Set},
+    convert::TryFrom,
+    fmt::Display,
+    path::PathBuf,
+    str::FromStr,
+};
+use structopt::*;
+use thiserror::*;
+use types::{smart_contracts, transactions};
+mod cts1;
+use concordium_contracts_common::Deserial;
+
+/// Name of the NFT smart contract from the example implementing the CTS1
+/// specification.
+const NFT_CONTRACT_NAME: &str = "CTS1-NFT";
+
+/// The NFT contract state for each address.
+/// Important: this structure is matching the NFT contract implementations and
+/// cannot be assumed for any CTS1 contract state, as the CTS1 does not restrict
+/// the state in anyway.
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+struct NFTContractAddressState {
+    /// The tokens owned by this address.
+    owned_tokens: Set<cts1::TokenIdVec>,
+    /// The address which are currently enabled as operators for this address.
+    operators:    Set<concordium_contracts_common::Address>,
+}
+
+/// Deserialization of AddressState.
+/// Important: this deserialization is matching the NFT contract implementations
+/// and cannot be assumed for any CTS1 contract state is serialized, as the CTS1
+/// does not restrict how to encode the state in anyway.
+impl Deserial for NFTContractAddressState {
+    fn deserial<R: concordium_contracts_common::Read>(
+        source: &mut R,
+    ) -> Result<Self, concordium_contracts_common::ParseError> {
+        let owned_tokens_length = u16::deserial(source)?;
+        let mut owned_tokens = Set::default();
+        for _ in 0..owned_tokens_length {
+            let k = cts1::TokenIdVec::deserial(source)?;
+            owned_tokens.insert(k);
+        }
```

Comment:
Use `deserial_set_no_length`, no need to duplicate this (and ignore duplicates on insertions).